Add lisflood class

SarahAlidoost opened the pull request.

SarahAlidoost committed:
a draft lisflood class

SarahAlidoost committed:
add lisflood to models init.py

SarahAlidoost committed:
add functions to create config files

SarahAlidoost committed:
refactor config generators, add lisvap_run

SarahAlidoost committed:
refactor everything

SarahAlidoost committed:
add check_work_dir func

sverhoeven committed:
Move scratch dir to global config + Move lisflood input files fields to setup argument as a dataclass

sverhoeven committed:
Started on unit test against https://github.com/ec-jrc/lisflood-usecases/tree/master/LF_lat_lon_UseCase

sverhoeven committed:
Parse NetCDF files + start/end from ForcingData

sverhoeven committed:
Use common parent of ESMValTool data files as forcing directory

No dependency on dataset anymore

sverhoeven committed:
Create dummy forcing output

sverhoeven committed:
Merge remote-tracking branch 'origin/master' into add_lisflood_class

sverhoeven committed:
Work on lisflood test

sverhoeven committed:
Use Path instead of str

sverhoeven committed:
Remove 1/0

sverhoeven committed:
Isort + added reindex_forcings + added forcing as dir support

sverhoeven committed:
Add singularity to CI

sverhoeven committed:
Dont create copies of items in CFG on module level

Hard to test with those

sverhoeven committed:
Also add image to run_lisvap

sverhoeven committed:
Forcing can be Path

sverhoeven committed:
Setup should return workdir when it is created for you

sverhoeven committed:
Use filename for lisvap config

sverhoeven committed:
One work_dir to rule them all

SarahAlidoost committed:
fix indentation in lisvap config

SarahAlidoost committed:
fix args, return exit_code, stdout, stderr

SarahAlidoost committed:
fix lisflood config name

SarahAlidoost committed:
fix No such file or directory: b'/scratch/shared/ewatercycle/lisflood_20210413_134355/lisflood_ERA-Interim_Meuse_pr_1990_1990.nc.nc'

sverhoeven committed:
Force midnight

sverhoeven committed:
Use exec

sverhoeven committed:
Make output_dir optional in reindex_forcings()

SarahAlidoost committed:
add example lisflood notebook

SarahAlidoost committed:
Merge branch 'add_lisflood_class' of github.com:eWaterCycle/ewatercycle into add_lisflood_class

SarahAlidoost committed:
fix error in model.update by fixing the forcing data or using data from comparison

sverhoeven committed:
Flatten config + fix mypy errors

sverhoeven committed:
Need grpc4bmi

SarahAlidoost committed:
add .expanduser() to forcing_dir

SarahAlidoost committed:
remove forcing generator from the notebook

SarahAlidoost committed:
revise docstrings and remove some of the TODO items

SarahAlidoost committed:
don't hardcode pathroot and model_mask

SarahAlidoost committed:
use parent dir of mask for mounting in containers

SarahAlidoost committed:
fix a bug

SarahAlidoost committed:
replace PathLike with Path for mask because of mypy error: PathLike does not have parent attr

SarahAlidoost committed:
Merge branch 'master' into add_lisflood_class

SarahAlidoost committed:
add .resolve() to forcing_dir to use absolute path

SarahAlidoost committed:
add get_value_as_xarray

SarahAlidoost committed:
test lisflood class with the notebook

sverhoeven committed:
Fix tests by mocking BmiClientSingularity

sverhoeven committed:
Started on test for run_lisvap

SarahAlidoost committed:
Merge branch 'master' into add_lisflood_class

SarahAlidoost committed:
add lisflood specific settings to config CFG

SarahAlidoost committed:
import CFG from ewatercycle, replace scratch_dir with output_dir

SarahAlidoost committed:
Merge branch 'master' into add_lisflood_class

SarahAlidoost committed:
add explanations about CFG to the notebook

Peter9192 commented on ewatercycle/models/lisflood.py:
```diff
@@ -0,0 +1,398 @@
+
+
+@dataclass
+class LisfloodParameterSet:
```
Nice, I like this solution

Peter9192 commented on ewatercycle/models/lisflood.py:
```diff
@@ -0,0 +1,398 @@
+
+# LISFLOOD settings file reference maps prefixes
+# MapsName: {prefix_name, prefix}
+MAPS_PREFIXES = {
```
Not sure I understand what this is needed for.

SarahAlidoost replied:
An explanation is added.

Peter9192 commented on ewatercycle/models/lisflood.py:
```diff
@@ -0,0 +1,398 @@
+              work_dir: PathLike = None) -> Tuple[PathLike, PathLike]:
+        """Configure model run
+
+        If forcing is missing evaporation files (e0, es0, et0) then the :py:meth:`run_lisvap` function should be run.
```
This instruction is a bit unclear. If my understanding is correct, then perhaps you could rephrase as:

```suggestion
        If evaporation files (e0, es0, et0) are not included in the forcing, then :py:meth:`run_lisvap` function should be called before setup.
```

Peter9192 commented on ewatercycle/models/lisflood.py:
```diff
@@ -0,0 +1,398 @@
+
+    # unable to subclass with more specialized arguments so ignore type
+    def setup(self,  # type: ignore
+              forcing: Union[ForcingData, PathLike],
```
Can forcing be part of the parameterset? Or should it always be passed seperately?

SarahAlidoost replied:
It can be, I added it as a TODO here.

sverhoeven replied:
Yep, that will make the forcing argument optional.

Peter9192 commented on ewatercycle/models/lisflood.py:
```diff
@@ -0,0 +1,398 @@
+    """eWaterCycle implementation of Lisflood hydrological model.
+
+    Attributes:
+        bmi (Bmi): Basic Modeling Interface object
+        work_dir (PathLike): Working directory for the model where it can read/write files
```
Why list (only) these and not e.g. `self.parameterset`?

SarahAlidoost replied:
I think because `parameterset` is given by users to the `setup()`. @sverhoeven what do you think?

sverhoeven replied:
Correct, anything given by user can be treated as private vars inside class, while things the class creates that are useful for user are listed here.

Peter9192 commented on ewatercycle/models/lisflood.py:
```diff
@@ -0,0 +1,398 @@
+
+    def _check_work_dir(self, work_dir):
+        """Check and optionally create the work directory."""
+        self.work_dir = _generate_workdir(work_dir)
```
Why is this a separate function? Isn't it a bit overkill to call `_check_work_dir` if the only thing it does is call `_generate_workdir`?

SarahAlidoost replied:
The `_generate_workdir` is also used in `reindex_forcings ` that is not part of the model class.

Peter9192 commented on ewatercycle/models/lisflood.py:
```diff
@@ -0,0 +1,398 @@
+        self.work_dir = _generate_workdir(work_dir)
+
+    def _check_forcing(self, forcing):
+        """"Check forcing argument."""
```
Can we make this docstring slightly more informative?

SarahAlidoost replied:
Sure, it is revised.

Peter9192 commented on ewatercycle/models/lisflood.py:
```diff
@@ -0,0 +1,398 @@
+                self.forcing_files[var_name] = forcing_file.name
+                # get start and end date of time dimension
+                # TODO converting numpy.datetime64 to datetime object is ugly, find better way
+                self.start = datetime.utcfromtimestamp(dataset.coords['time'][0].values.astype('O') / 1e9)
+                self.end = datetime.utcfromtimestamp(dataset.coords['time'][-1].values.astype('O') / 1e9)
```
Maybe something to think about: do we want the forcing data to determine the start/end time, or do we want the used to set the start/end time independently of the forcing?

SarahAlidoost replied:
Good question. If the forcing is the only temporal data given to lisflood, it makes sense that the start/end time of forcing is used in configuration and thus in initialization. Users can choose any start/end time (in forcing time range) while runs the model.

sverhoeven replied:
The forcing gives a maximum range. A user can use update() to stop early, but now can not start later. Supplying a start date to setup could help, but not in scope for current use cases.

SarahAlidoost replied:
We added a TODO item in setup().

Peter9192 commented on ewatercycle/models/lisflood.py:
```diff
@@ -0,0 +1,398 @@
+    return work_dir
+
+
+class XmlConfig(AbstractConfig):
```
Should we move this to `ewatercycle/parametersetdb/config` so it can be reused by other (future) model classes as well?

SarahAlidoost replied:
We (@sverhoeven and I) discussed this. `source` is a path in this class whereas it is an URL in other classes in `ewatercycle/parametersetdb/config`.

Peter9192 replied:
Oh right, yes, I stumbled upon that as well. Perhaps we can try to make that more flexible in `parametersetdb`?

sverhoeven replied:
The LisFlood model is the only model that uses XML, when we add a another model that uses XML then we can move it over

Peter9192 replied:
Created an issue: https://github.com/eWaterCycle/ewatercycle/issues/58

Peter9192 commented on ewatercycle/models/lisflood.py:
```diff
@@ -0,0 +1,398 @@
+                    textvar.set('value', f"$(PathOut)/$({prefix['name']})")
+
+        # Write to new setting file
+        lisflood_file = f"{self.work_dir}/lisflood_setting_{timestamp}.xml"
```
Why do we need a timestamp in the settings filename?

SarahAlidoost replied:
Good point, it is removed.

Peter9192 commented on ewatercycle/models/lisflood.py:
```diff
@@ -0,0 +1,398 @@
+        """Run lisvap to generate evaporation input files
+
+        Args:
+            forcing: Path to re-indexed forcings
```
What do you mean by "re-indexed"?

SarahAlidoost replied:
it is removed.

Peter9192 commented on ewatercycle/models/lisflood.py:
```diff
@@ -0,0 +1,398 @@
+        cfg.save(lisflood_file)
+        return lisflood_file
+
+    def _create_lisvap_config(self) -> str:
```
There seems to be quite a bit of duplication here. Is there a nicer way to use shared config settings between lisflood and lisvap? Or are they too different?

SarahAlidoost replied:
I don't think so. lisflood and lisfvap are two separate modules. I think that it is more clear if they have two separate config files, for more info on lisflood documentation, see https://ec-jrc.github.io/lisflood-model/.

Peter9192 commented on ewatercycle/models/lisflood.py:
```diff
@@ -0,0 +1,398 @@
+        return da
+
+    @property
+    def parameters(self) -> Iterable[Tuple[str, Any]]:
```
I don't really understand why these would be the 'configurable' parameters. Wouldn't it make more sense if it listed the options in the lisflood config file?

SarahAlidoost replied:
I agree, @sverhoeven what do you think?

sverhoeven replied:
Now setup and parameters are sides of the same coin. So any change we make here should also be done in setup().
The forcing dir and working dir become part of the config file.

Which arguments should the setup have?  In wflow it is very dynamic but luckily 2 levels deep. In Lisflood there is a more complicated config with differently shaped sections and nestings.

Peter9192 replied:
I'm not sure either. I just think the philosophy of PyMT is that these parameters are the things you'd like to change about the model configuration. So more like the default values from the [tables here](https://ec-jrc.github.io/lisflood-model/3_01_optLISFLOOD_overview/). But I agree that there is probably no neat way to do that right now. Perhaps one way would be to do it through the ParameterSet class you created. 

I think I'd prefer a NotImplementedError over the current solution, because I'm not very happy about e.g. having to pass a work_dir on setup and this feels like advertising that. 

Also note that it's currently not the same as the arguments to `model.setup`: parameterset is missing.

Peter9192 commented on ewatercycle/models/lisflood.py:
```diff
@@ -0,0 +1,398 @@
+        if self.forcing_dir and self.work_dir:
+            return {
+                'forcing_dir': self.forcing_dir,
+                'work_dir': self.work_dir,
```
I have a bit of an internal conflict about the naming of `work_dir` (which is what grpc4bmi uses) and `cfg_dir` (which is what PyMT uses). I'm inclined to go with the latter for everything that the user interacts with, to stay in line with PyMT.

SarahAlidoost replied:
I agree. In [PyMT usage](https://pymt.readthedocs.io/en/latest/usage.html), the variable is named as `cfg_dir`. However, it is referred to **run directory** in the text. Let's make an issue for this and discuss it in a design session.

Peter9192 commented on ewatercycle/models/lisflood.py:
```diff
@@ -0,0 +1,398 @@
+        return []
+
+
+def reindex_forcings(mask_map: PathLike, forcing: ForcingData, output_dir: PathLike = None) -> PathLike:
```
I'm not sure I understand this. If this is necessary, doesn't that mean that the forcing is just not 'ready' yet as it comes out of ESMValTool?

SarahAlidoost replied:
This is the last part in the lisflood preprocessing notebook in comparison. The grid of forcing data generated by ESMValTool does not match the lisflood grid. The lisflood grid is given by a `model_mask` file that is available in its parameterset.

Peter9192 commented on ewatercycle/models/lisflood.py:
```diff
@@ -0,0 +1,398 @@
+        """"Check forcing argument."""
+
+        # TODO check if mask has same grid as forcing files,
+        # if not warn users to run reindex_forcings
```
Should the user be concerned with this or should it be done automatically?

SarahAlidoost replied:
I think users.

Peter9192 commented on ewatercycle/models/lisflood.py:
```diff
@@ -0,0 +1,398 @@
+    return output_dir
+
+
+def _generate_workdir(work_dir: PathLike = None) -> PathLike:
```
Why is this a function rather than a method?

SarahAlidoost replied:
The `_generate_workdir` is also used in `reindex_forcings` that is not part of the model class.

Peter9192 replied:
See my new comment on this usage from withing `reindex_forcings`. It seems it's not doing anything. I still don't understand.

SarahAlidoost replied:
right, the function `_check_work_dir` is removed now.

Peter9192 commented on setup.py:
```diff
@@ -21,6 +21,7 @@
     install_requires=[
         'basic_modeling_interface',
         'esmvaltool',
+        'grpc4bmi>=0.2.12,<0.3',
```
I think this is already part of master, no?

SarahAlidoost replied:
Good catch, yes, it is sorted now.

Peter9192 commented on tests/models/test_lisflood.py:
```diff
@@ -0,0 +1,111 @@
+from ewatercycle.models.lisflood import Lisflood, LisfloodParameterSet
+
+
+@pytest.fixture
```
For my understanding, could you explain what we are really testing here? Since we're mostly testing against mocks and fixtures..

SarahAlidoost replied:
@sverhoeven could you please reply to this question? thanks.

sverhoeven replied:
The only thing not real is the bmi object to make tests quick. This test exercises setup() method, but does not check content of config file.

As we switched from unit testing against to lisflood use case (https://github.com/ec-jrc/lisflood-usecases/tree/master/LF_lat_lon_UseCase) to using the notebook for testing. We stopped further developments here.

Peter9192 reviewed (changes requested):
Hey @SarahAlidoost really impressive work here! The lisflood class seems to be quite a bit more involved than wflow! 

Please see my other comments below and let me know if anything is unclear.

Peter9192 commented:
I left some comments via reviewnb as well, but I don't see them show up here. Please see: https://app.reviewnb.com/eWaterCycle/ewatercycle/blob/add_lisflood_class/examples%2Flisflood.ipynb%2F/discussion

SarahAlidoost committed:
add explanations about MAPS_PREFIXES, revise docstring in _check_forcing, remove timestamp from config name, add a TODO in setup

SarahAlidoost committed:
remove duplication and sort packages in setup.py

SarahAlidoost committed:
revise docstring in setup

Co-authored-by: Peter Kalverla <peter.kalverla@gmx.com>

SarahAlidoost committed:
import CFG, use yaml code block, add  model.get_value_as_xarray and plot to the notebook

SarahAlidoost commented:
> Hey @SarahAlidoost really impressive work here! The lisflood class seems to be quite a bit more involved than wflow!
> 
> Please see my other comments below and let me know if anything is unclear.

@Peter9192 thank you for reviewing and helpful comments. For some of your questions on code and reviewnb, I asked @sverhoeven to help with answers. I have to re-run the notebook and push the results. I will inform you when it is done.

SarahAlidoost committed:
add TODO for start time argument in setup()

SarahAlidoost committed:
fix bug in get_value__as_xarray

SarahAlidoost committed:
add supress warnings, add plot, re-run notebook

Peter9192 commented on ewatercycle/models/lisflood.py:
```diff
@@ -0,0 +1,400 @@
+    Returns:
+        Output dir with re-indexed files.
+    """
+    output_dir = _generate_workdir(output_dir)
```
I don't understand. Here you pass `output_dir` as the value for the `work_dir` argument. But if I read `_generate_workdir` correctly, it will then just return output_dir, right? So it will not do anything.

SarahAlidoost replied:
right, the function `_check_work_dir` is now removed.

SarahAlidoost commented:
> Hey @SarahAlidoost really impressive work here! The lisflood class seems to be quite a bit more involved than wflow!
> 
> Please see my other comments below and let me know if anything is unclear.

Hey @Peter9192, we (@sverhoeven and I) addressed your comments. Please have another look, thanks.

SarahAlidoost committed:
remove  _check_work_dir func

SarahAlidoost committed:
add a todo for model.parameters link to an issue

SarahAlidoost committed:
Merge branch 'master' into add_lisflood_class

Peter9192 commented:
Thansk @SarahAlidoost and @sverhoeven , great work!
